examples/hello/lib/hello/app.rb

```diff
@@ -0,0 +1,66 @@
+require "hello/actions"
+
+module Hello
+  class App
+    class Router
+      class NotFound < StandardError; end
+
+      attr_reader :routes
+
+      def initialize(&block)
+        @routes = {}
+        instance_eval(&block)
+      end
+
+      def resource(name, at:, &block)
+        @this_path = at
+        routes[@this_path] = {}
+        instance_eval(&block)
+      end
+
+      def on(req_method, call:)
+        routes[@this_path][req_method] = call
+      end
+
+      def action_for(path, req_method)
+        routes.fetch(path).fetch(req_method)
+      rescue KeyError
+        raise NotFound.new("Can't find route for #{req_method} #{path}")
+      end
+    end
+
+    def call(env)
+      router = Router.new do
+        resource "hello", at: "/hello" do
+          on "GET", call: "say_hello"
+          on "HEAD", call: "do_nothing"
+        end
+
+        resource "world", at: "/world" do
+          on "DELETE", call: "say_goodbye"
+        end
+      end
+
+      path = env.fetch("PATH_INFO")
+      req_method = env.fetch("REQUEST_METHOD")
+
+      begin
+        action_name = router.action_for(path, req_method)
+        action = actions.send(action_name)
+        body = action.call
```

Comment:
I referred to the router as a configuration object. Maybe it should explicitly be an immutable value object? Maybe it should be called "routes" instead of "router"?

```ruby
Dispatcher.new(routing, actions).dispatch(request)
```

```ruby
def dispatch(request)
  route = routes.get(request.path, request.method)
  action = actions.get(route.action_name)
  action.call
end
```

...which is essentially the same procedure, but the naming doesn't make it feel as wrong.